File: support-frontend/assets/helpers/internationalisation/currency.js
// @flow

// ----- Imports ----- //

import { getQueryParameter } from 'helpers/url';

import {
  type CountryGroup,
  type CountryGroupId,
  countryGroups,
} from './countryGroup';

// ----- Types ----- //

export type IsoCurrency =
  | 'GBP'
  | 'USD'
  | 'AUD'
  | 'EUR'
  | 'NZD'
  | 'CAD'
  | 'SEK';

export type Currency = {|
  glyph: string,
  extendedGlyph: string,
|};

export type SpokenCurrency = {|
  singular: string,
  plural: string,
|};


// ----- Config ----- //

const currencies: {
  [IsoCurrency]: Currency,
} = {
  GBP: {
    glyph: '£',
    extendedGlyph: '£',
  },
  USD: {
    glyph: '$',
    extendedGlyph: 'US$',
  },
  AUD: {
    glyph: '$',
    extendedGlyph: 'AU$',
  },
  EUR: {
    glyph: '€',
    extendedGlyph: '€',
  },
  NZD: {
    glyph: '$',
    extendedGlyph: 'NZ$',
  },
  CAD: {
    glyph: '$',
    extendedGlyph: 'CA$',
  },
  SEK: {
    glyph: 'kr',
    extendedGlyph: 'kr',
  },
};

const spokenCurrencies: {
  [IsoCurrency]: SpokenCurrency,
} = {
  GBP: {
    singular: 'pound',
    plural: 'pounds',
  },
  USD: {
    singular: 'dollar',
    plural: 'dollars',
  },
  AUD: {
    singular: 'dollar',
    plural: 'dollars',
  },
  EUR: {
    singular: 'euro',
    plural: 'euros',
  },
  NZD: {
    singular: 'dollar',
    plural: 'dollars',
  },
  CAD: {
    singular: 'dollar',
    plural: 'dollars',
  },
  SEK: {
    singular: 'krona',
    plural: 'kronor',
  },
};


// ----- Functions ----- //

function fromCountryGroupId(countryGroupId: CountryGroupId): ?IsoCurrency {
  const countryGroup: ?CountryGroup = countryGroups[countryGroupId];

  if (!countryGroup) {
    return null;
  }

  return countryGroup.currency;
}

function fromString(s: string): ?IsoCurrency {
  switch (s.toLowerCase()) {
    case 'gbp': return 'GBP';
    case 'usd': return 'USD';
    case 'aud': return 'AUD';
    case 'eur': return 'EUR';
    case 'nzd': return 'NZD';
    case 'cad': return 'CAD';
    case 'sek': return 'SEK';
    default: return null;
  }
}

function fromQueryParameter(): ?IsoCurrency {
  const currency = getQueryParameter('currency');
  if (currency) {
    return fromString(currency);
  }
  return null;
}

function detect(countryGroup: CountryGroupId): IsoCurrency {
  return fromQueryParameter() || fromCountryGroupId(countryGroup) || 'GBP';
}

const glyph = (c: IsoCurrency): string => currencies[c].glyph;
const extendedGlyph = (c: IsoCurrency): string => currencies[c].extendedGlyph;

// ----- Exports ----- //

export {
  detect,
  spokenCurrencies,
  fromCountryGroupId,
  currencies,
  glyph,
  extendedGlyph,
};

File: support-internationalisation/src/main/scala/com/gu/i18n/Currency.scala
package com.gu.i18n

sealed trait Currency {
  def prefix: Option[String] = None
  def glyph: String
  def identifier: String = prefix.getOrElse("") + glyph
  def iso: String
}

object Currency {
  val all = List(GBP, USD, AUD, CAD, EUR, NZD, SEK, CHF, NOK, DKK)

  def fromString(s: String): Option[Currency] = all.find(_.iso == s)

  case object GBP extends Currency {
    override def glyph: String = "£"
    override def iso: String = "GBP"
  }
  case object USD extends Currency {
    override def glyph: String = "$"
    override def prefix: Option[String] = Some("US")
    override def iso: String = "USD"
  }
  case object AUD extends Currency {
    override def glyph: String = "$"
    override def prefix: Option[String] = Some("AU")
    override def iso: String = "AUD"
  }
  case object CAD extends Currency {
    override def glyph: String = "$"
    override def prefix: Option[String] = Some("CA")
    override def iso: String = "CAD"
  }
  case object EUR extends Currency {
    override def glyph: String = "€"
    override def iso: String = "EUR"
  }
  case object NZD extends Currency {
    override def glyph: String = "$"
    override def prefix: Option[String] = Some("NZ")
    override def iso: String = "NZD"
  }
  case object SEK extends Currency {
    override def glyph: String = "kr"
    override def prefix: Option[String] = Some("SE")
    override def iso: String = "SEK"
  }
  case object CHF extends Currency {
    override def glyph: String = "fr"
    override def prefix: Option[String] = Some("CH")
    override def iso: String = "CHF"
  }
  case object NOK extends Currency {
    override def glyph: String = "kr"
    override def prefix: Option[String] = Some("NO")
    override def iso: String = "NOK"
  }
  case object DKK extends Currency {
    override def glyph: String = "kr."
    override def prefix: Option[String] = Some("DK")
    override def iso: String = "DKK"
  }
}
